### PR TITLE
MAGESHIP-21-develop Add track_url to DB & API and send via Customer S…

### PR DESCRIPTION
**Pull request by RhodriOwainDavies**
**Please beware this is critical for Magento Shipping (to be released in v2.2.2 - discuss with @joni-jones @orlangur & Nadia Kalinovich**

[Linked to #10799](https://github.com/magento/magento2/pull/10799)

Forked from `develop` branch, merge target branch `develop` as per @orlangur 's request.

### Description
Add a new field sales_shipment_track.track_url.
Expose this field via API.
Including this URL as a hyperlink in the Shipment Sent Email template.

### Feature
To support impending Magento Shipping feature request of inserting Carrier Specific URL in email template to Customer

### Manual testing scenarios
1. API Test
POST /V1/order/{orderId}/ship
Pass in tracking_url as an extension attribute (in same manner as tracking_number)
2. UI Test
In the admin section, view a Shipment which already has a trackinging URL set (via API in Test #1).  Click Send Tracking Email.  Check email, Tracking Number should be a hyperlink to tracking_url

### Contribution checklist
 - [X] Pull request has a meaningful description of its purpose
 - [X] All commits are accompanied by meaningful commit messages
 - [X] All new or changed code is covered with unit/integration tests (if applicable)
 - [X] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

File: app/code/Magento/Sales/composer.json [hunk 1]
```diff
@@ -32,7 +32,7 @@
         "magento/module-sales-sample-data": "Sample Data version:100.3.*"
     },
     "type": "magento2-module",
-    "version": "100.3.0-dev",
+    "version": "100.3.1-dev",
```
Comment (vrann) [inline app/code/Magento/Sales/composer.json]: This change is not needed, the version of the extension will be bumped automatically during the build depending on the severity of the changes

**Review by vrann — CHANGES_REQUESTED** [review]

**Review by vrann — COMMENTED** [review]
TrackUrl interface is added to the system but never referenced neither from the interfaces in Api namespace nor from the extensions attributes. How is it expected to be exposed through the API?

I tried to run API query adding track_url to the body, but it resulted in decoding error, because Property `TrackUrl` does not have corresponding setter in class `Magento\Sales\Api\Data\ShipmentTrackCreationInterface`

Please advise.

Request:
```
{
  "items": [
    {
      "extension_attributes":[],
      "order_item_id":3,
      "qty":1
    
    },
    {
      "extension_attributes":[],
      "order_item_id":4,
      "qty":1
    }
  ],
  "notify":true,
  "appendComment":true,
  "comment":{
    "extension_attributes":[],
    "comment":"Shipped via Shipment Provider",
    "is_visible_on_front":0
  },
  "tracks":[
    {
      "extension_attributes":[],
      "track_number":"1234567890",
      "track_url": "http://fedex.com/122321212",
      "title":"fedex",
      "carrier_code":
      "fedex"
    }
  ],
  "packages":[
    {
      "extension_attributes":[]
    }
  ],
  "arguments":{
    "extension_attributes":[]
  }
}
```

File: app/code/Magento/Sales/Api/Data/ShipmentTrackInterface.php [hunk 2]
```diff
@@ -46,6 +46,10 @@
      */
     const TRACK_NUMBER = 'track_number';
     /*
+     * Track URL.
+     */
+    const TRACK_URL = 'track_url';
```
Comment (vrann) [inline app/code/Magento/Sales/Api/Data/ShipmentTrackInterface.php]: this constant does not have corresponding getter and setter in this interface and should not reside here, rather in the interface which has setter for it.

**Review by vrann — CHANGES_REQUESTED** [review]

File: app/code/Magento/Sales/composer.json (hunk 1 above)
Comment (RhodriOwainDavies) [reply app/code/Magento/Sales/composer.json]: Updated

**Comment by RhodriOwainDavies** [comment]
Hi @vrann 

Thanks for your review, I've addressed all your comments.  Can you please test again and let me know.  If the changes are successful, I'll update #10977 with the same changes.

Cheers,

Rhodri D

File: composer.json [hunk 3]
```diff
@@ -160,7 +160,7 @@
         "magento/module-robots": "100.3.0-dev",
         "magento/module-rss": "100.3.0-dev",
         "magento/module-rule": "100.3.0-dev",
-        "magento/module-sales": "100.3.0-dev",
```
Comment (vrann) [inline composer.json:163]: This change is not needed, it will be automatically set by the packaging tool

**Review by vrann — CHANGES_REQUESTED** [review]
I was expecting to see tracking URL in the admin backend and in the customer account on the frontend, but neither of those places has it. Is it just sent via email? Please add them to order shipments and to the customer account.
screen shot 2017-10-05 at 4 54 14 pm
screen shot 2017-10-05 at 4 53 27 pm

**Comment by vrann** [comment]
@RhodriOwainDavies discussed with @antonkril today, this change cannot be released in 2.2 as backward incompatible. Please re-create the PR for `develop` branch

**Pull request closed without merging**